runtime/src/iree/base/tree.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_BASE_TREE_H
+#define IREE_BASE_TREE_H
+
+#include "iree/base/allocator.h"
+#include "iree/base/config.h"
+#include "stdint.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+typedef struct iree_tree_node_t iree_tree_node_t;
+typedef struct iree_tree_t iree_tree_t;
+
+typedef enum iree_tree_walk_type_e {
+  IREE_TREE_WALK_PREORDER,
+  IREE_TREE_WALK_INORDER,
+  IREE_TREE_WALK_POSTORDER,
+} iree_tree_walk_type_e;
+
+typedef struct iree_tree_node_t iree_tree_node_t;
+typedef struct iree_tree_node_t {
+  bool red;
+  iree_tree_node_t* left;
+  iree_tree_node_t* right;
+  iree_tree_node_t* parent;
+  iree_host_size_t key;
```

Comment:
The key is simply an integer value. We could make it pointer-sized so that we can always insert pointers, but the intention here was to just have it be integral. I am happy to change it if we want.